### ISSUE TITLE
Local Pickup instance id not being copied to order

### ISSUE THREAD

**Issue by senadir** [woothemes/woocommerce#47060]
When placing an order using Local Pickup, the instance id is not being copied correctly to the order, and it's always 0.

In Local Pickup, each location (rate, instance) has an instance id that matches its index on the Local Pickup settings page.

When placing an order, the instance id is correct, after being placed, the instance id is set to 0.

### To reproduce

- Have 2 Local Pickup locations.
- In Checkout, select the second location.
- In console, run the following code, notice that it returns `pickup_location:1`
```js
wp.data.select( 'wc/store/cart' ).getCartData().shippingRates[0].shipping_rates.find(rate => rate.selected).rate_id
```
- Place the order, note its ID.
- From WP-CLI, enter shell mode via `wp shell`.
- Run the following code:
```php
$order = wc_get_order( id-here );
reset($order->get_shipping_methods())->get_instance_id();
```
- You will get 0, it should be 1.

### Why is this important

Apart from the instance id, there's no reliable way to know which location was picked, as string can be modified by translation systems, so instance id need to be copied like it happens for regular shipping.

**Comment by senadir** [comment]
Local Pickup is a non-zone shipping method, it's worth confirming if this is a bug in non-zone based shipping methods, or just local pickup.

**Comment by mikejolley** [comment]
@senadir It doesn't look like Local Pickup uses instance IDs. You see here it is index based from the location array:

https://github.com/woocommerce/woocommerce/blob/da18c925b9c9ae05199198662f80370a30b9f77e/plugins/woocommerce/src/Blocks/Shipping/PickupLocation.php#L97-L103

The method ID is "pickup_location". Any location has the same method ID of "pickup_location".

Zone based shipping methods are different. With zones, you create an instance of a shipping method e.g. "flat_rate", and then its stored as an actual entity in the database as seen here:

![Screenshot 2024-08-07 at 13 44 40](https://github.com/user-attachments/assets/726c0207-e5ff-4ba2-b0e0-bdad27721f40)

Note the `instance_id` column in the table associated with each shipping method.

Since local pickup is not part of the zones system, `instance_id` is always `0`. Meta on the shipping line item stores what option was selected in terms of location.

**Comment by senadir** [comment]
If `get_instance_id` is not the best way to get the value, what would be?  `get_id`?

**Comment by mikejolley** [comment]
@senadir since there is no entity per location row (its all serialised) you'd use the line item meta data `pickup_location` which contains the text based string of the location name.

**Comment by mikejolley** [comment]
I will remove my assignment for now and marks this as needing further discussion. If there is a need for the locations themselves to become some form of entity in the database this is a much larger task. cc @opr

**Comment by pierorocca** [comment]
This needs engineering input before we can confirm the final disposition.

@senadir, I'm supportive of your recommendation here to resolve this gap. Do you see a path forward that would make this "ready for engineering"?

**Comment by senadir** [comment]
I think we can look into introducing stable ID for pickup locations (based on numbers on addresses + country, state, location in array as fallback). This would be for newly added pickup locations. Each time a location is updated, its id should be updated.
